Client/Assets/CustomSearchHelper.js

```diff
@@ -0,0 +1,44 @@
+//xcode is soooo garbage with js. whyyyyyyyyy
+
+if (!window.__firefox__) {
+  window.__firefox__ = {};
+}
+
+(function() {
+  window.__firefox__.isActiveElementSearchField = function() {
+    var input = document.activeElement;
+    if (input.tagName !== "INPUT") return "false"
+    var form = input.form
+    if (!form || form.method != 'get') return "false"
+    return "true"
+  }
+
+  window.__firefox__.searchQueryForField = function() {
+    var input = document.activeElement;
+    if (input.tagName !== "INPUT") return {}
+    var form = input.form
+    if (!form || form.method != 'get') return {}
+    
+    var inputs = form.getElementsByTagName('input');
+    inputs = Array.prototype.slice.call(inputs,0);
+    var params = inputs.map(function(element){
+      if (element.name == input.name) return [element.name,"{searchTerms}"].join('=');
+      return [element.name,element.value].join('=');
+    });
+
+    var selectFields = form.getElementsByTagName('select');
+    selectFields = Array.prototype.slice.call(selectFields,0);
+    var selectParams = selectFields.map(function(e){
+      e.options[e.selectedIndex].value;
+      return [e.name,e.options[e.selectedIndex].value].join('=')
+    });
+    params = params.concat(selectParams)
+    if (!form.action) return {} //an invalid form.
+    var url = [form.action,params.join('&')].join('?')
+
+    var iconElement = document.head.querySelector("link[rel~='icon']")
```

Comment:
Each `Tab` object in Swift has a `favicons` array containing the set of favicons for the page. Those are provided by `Favicons.js`, which has a couple other selectors, too (namely, Apple touch icons). I imagine we'd be able to just reuse that when receiving the message instead of having to reimplement favicon logic here.